### PR TITLE
Refactor metrics handler

### PR DESCRIPTION
**Pull request by will-cromar**
Large refactor of metrics handler. As discussed internally, we're removing the intermediate queue of test events such that one test run will map to one pubsub message ([except in edge cases](https://cloud.google.com/pubsub/docs/faq)) and one Function call (unless there's a transient error). Test events will be published from Deployment in cluster, rather than from an initContainer in each job (coming soon in another PR).

- Explicitly define metric collection configuration with a protocol buffer schema.
- Handle a single event at a time in the metrics handler to simplify code
- Decouple the metrics handler from the GKE clusters that run the tests.
- Re-organize `main.py` and `metrics.py` into multiple files.
  - Most stateful interactions will stay in `main.py`.
  - Factor out interactions with BigQuery to `bigquery_client.py`.
    - Define BigQuery schema with the new Python `@dataclass`es
    - Safely template queries with query parameters.
  - Move metric collection logic from `main.py` and `metrics.py` into `collectors/` folder. Each collector in that folder implements a common interface and corresponds to a specific metric source: TensorBoard, PerfZero, and the test event itself. This will make it easier to implement new metric sources.
- Re-implement AlertHandler as a log handler that catches error logs and templates them into an e-mail.
  - Safely template e-mails with Jinja instead of string.format.
- Use [`functions-framework`](https://github.com/GoogleCloudPlatform/functions-framework-python) to enable local testing and portability to e.g. Cloud Run (instructions coming soon)
- [Increase test coverage](https://github.com/GoogleCloudPlatform/ml-testing-accelerators/pull/348#issuecomment-742123103) 🎊

### CODE REVIEW

**Review by zcain117 — COMMENTED** [review]
Sending a few comments but haven't finished reading all the way yet

File: metrics_v2/handler/collectors/base.py [hunk 1]
```diff
@@ -0,0 +1,138 @@
...
+          self._event.benchmark_id),
+      metric_key=metric_key,
+      min_time=max(
+          min_timestamp.ToDatetime(),
+          self._event.start_time.ToDatetime() - time_window.ToTimedelta())
+    )
+    
+    return [row.metric_value for row in history_rows]
+
+  def compute_bounds(self, metric_key, assertion) -> utils.Bounds:
```
[the remark this answers is not in the document]
Comment (will-cromar) [reply metrics_v2/handler/collectors/base.py:61]: Done. Added type hints and updated docstrings here and everywhere else.

File: metrics_v2/handler/collectors/base.py [hunk 2]
```diff
@@ -0,0 +1,138 @@
...
+
+      mean = np.mean(values)
+      stddev = np.std(values)
+      c = assertion.std_devs_from_mean.comparison
+      if c in (metrics_pb2.Assertion.LESS, metrics_pb2.Assertion.WITHIN):
+        upper_bound = mean + (stddev * assertion.std_devs_from_mean.std_devs)
+      if c in (metrics_pb2.Assertion.GREATER, metrics_pb2.Assertion.WITHIN):
+        lower_bound = mean - (stddev * assertion.std_devs_from_mean.std_devs)
+    elif assertion_type == 'percent_difference':
+      target_type = assertion.percent_difference.WhichOneof('target_type')
```
[the remark this answers is not in the document]
Comment (will-cromar) [reply metrics_v2/handler/collectors/base.py:105]: I think it's possible for both values to be unset. I added a an `else` block to catch this case.

File: metrics_v2/handler/collectors/base.py [hunk 3]
```diff
@@ -0,0 +1,138 @@
...
+      elif target_type == 'value':
+        target = assertion.percent_difference.target
+
+      c = assertion.percent_difference.comparison
+      if c in (metrics_pb2.Assertion.LESS, metrics_pb2.Assertion.WITHIN):
+        upper_bound = target + (assertion.percent_difference.percent * target)
+      if c in (metrics_pb2.Assertion.GREATER, metrics_pb2.Assertion.WITHIN):
+        lower_bound = target - (assertion.percent_difference.percent * target)
+    
+    if upper_bound == math.inf and lower_bound == math.inf:
```
[the remark this answers is not in the document]
Comment (will-cromar) [reply metrics_v2/handler/collectors/base.py:129]: 🤦 fixed

File: metrics_v2/watcher/job_watcher.py [hunk 4]
```diff
@@ -0,0 +1,102 @@
+import re
```
Comment (will-cromar) [inline metrics_v2/watcher/job_watcher.py:1]: This PR is already very large. I will move the `watcher/` code to a new PR before I finish this draft.

File: metrics_v2/handler/collectors/base_test.py [hunk 5]
```diff
@@ -0,0 +1,79 @@
...
+    ('less_or_equal', 'LESS', 1, (-math.inf, 4.414), True),
+    ('greater', 'GREATER', 1, (1.586, math.inf)),
+    ('greater_or_equal', 'GREATER', 1, (1.586, math.inf), True),
+    ('less_2_dev', 'LESS', 2, (-math.inf, 5.828)),
+    ('within', 'WITHIN', 1, (1.586, 4.414)),
+  )
+  def test_compute_bounds_stddevs_from_mean(
+      self, comparison, std_devs, expected_bounds, inclusive=False):
+    # mean = 3, stddev = ~1.414
+    value_history = list(range(1, 6))
```
[the remark this answers is not in the document]
Comment (will-cromar) [reply metrics_v2/handler/collectors/base_test.py]: Oops, for some reason I thought range was a generator. Fixed.

[the remark this answers is not in the document]
Comment (zcain117) [reply metrics_v2/handler/collectors/base_test.py]: `xrange` was a generator in python2

File: metrics_v2/handler/collectors/literal_collector.py [hunk 6]
```diff
@@ -0,0 +1,23 @@
...
+from handler.collectors import base
+
+class LiteralCollector(base.BaseCollector):
+  def read_metrics_and_assertions(self):
+    for metric_key, assertion in self._source.assertions.items():
+      try:
+        # TODO: implement lookups in nested fields
+        raw_value = getattr(self._event, metric_key)
+      except AttributeError:
+        logging.error('Attribute %s not found in event', metric_key)
```
[the remark this answers is not in the document]
Comment (will-cromar) [reply metrics_v2/handler/collectors/literal_collector.py:12]: That's correct. This alert will get rolled up with other error-level messages, and this will only print if the user specified that a value should be present.

File: metrics_v2/handler/collectors/perfzero_collector.py [hunk 7]
```diff
@@ -0,0 +1,35 @@
...
+class PerfZeroCollector(base.BaseCollector):
+  def read_metrics_and_assertions(self):
+    """Yields aggregated metrics from a PerfZero summary file.
+
+    Values from process_info are prefixed with `process_info`.
+    """
+    glob_pattern = os.path.join(self.output_path, '*', 'perfzero_summary.json')
+    file_matches = tf.io.gfile.glob(glob_pattern)
+    if not file_matches:
+      logging.info('No perfzero summary found.')
```
[the remark this answers is not in the document]
Comment (will-cromar) [reply metrics_v2/handler/collectors/perfzero_collector.py]: Good idea. Changed to error.

File: metrics_v2/handler/collectors/perfzero_collector.py [hunk 8]
```diff
@@ -0,0 +1,35 @@
...
+      return
+
+    file_path = file_matches[0]
+    with tf.io.gfile.GFile(file_path) as f:
+      summary = json.loads(f.read())
+      timestamp = summary["execution_timestamp"]
```
[the remark this answers is not in the document]
Comment (will-cromar) [reply metrics_v2/handler/collectors/perfzero_collector.py]: For a partial run, I think this entire file would be missing, which caught above. I added a catches for KeyError and JSONDecodeError that print the file content and exception.

File: metrics_v2/handler/collectors/tensorboard_collector.py [hunk 9]
```diff
@@ -0,0 +1,132 @@
...
+
+# Represents a single summary scalar and the time it was recorded.
+TensorBoardScalar = collections.namedtuple('TensorBoardScalar', ['metric_value', 'wall_time'])
+
+class TensorBoardCollector(base.BaseCollector):
+  def __init__(self, *args, **kwargs):
+    super().__init__(*args, **kwargs)
+
+  def _prefixed_tag(self, tag: str) -> str:
+    if self._source.merge_runs and run != '.':
```
[the remark this answers is not in the document]
Comment (will-cromar) [reply metrics_v2/handler/collectors/tensorboard_collector.py]: Yes, it should be. I also fixed the tests to catch that problem.

File: metrics_v2/handler/collectors/tensorboard_collector.py [hunk 10]
```diff
@@ -0,0 +1,132 @@
...
+      return '/'.join((run, tag))
+
+    return tag
+
+  def _include_tag(self, tag: str) -> bool:
+    tag_path = pathlib.PurePath(tag)
+    return any(tag == a.tag for a in self._source.aggregate_assertions) or (
+        any(tag_path.match(x.tag_pattern) for x in self._source.include_tags)
```
[the remark this answers is not in the document]
Comment (will-cromar) [reply metrics_v2/handler/collectors/tensorboard_collector.py:36]: Thanks! I added a docstring to clarify the rule.

File: metrics_v2/handler/collectors/tensorboard_collector.py [hunk 11]
```diff
@@ -0,0 +1,132 @@
...
+        strategy_to_assertion = {
+          aggregate_assertion.strategy: aggregate_assertion.assertion
+          for aggregate_assertion in tag_to_aggregate_assertions[tag]
+        }
+        strategies.update(strategy_to_assertion.keys())
+      else:
+        strategy_to_assertion = {}
+
+      for strategy in strategies:
+        assertion = strategy_to_assertion.get(strategy)
```
[the remark this answers is not in the document]
Comment (will-cromar) [reply metrics_v2/handler/collectors/tensorboard_collector.py:135]: Assertion being `None` is valid here. Compute bounds will return `[-inf, inf]`

**Comment by will-cromar** [comment]
I'm mostly ready to call this PR done. There are a few more rough edges that I want to smooth out, but I will address those when I add the event publisher and integration tests. I will update my PR description to be more complete and mark this ready for review.

Coverage metrics for these files:

```
Name                                               Stmts   Miss  Cover   Missing
--------------------------------------------------------------------------------
handler/__init__.py                                    0      0   100%
handler/alerts.py                                     25      0   100%
handler/alerts_test.py                                61      0   100%
handler/bigquery_client.py                           100     49    51%   57, 90, 102-114, 125-133, 144, 148-150, 159, 163, 167-178, 188-219, 239-266
handler/collectors/__init__.py                         3      0   100%
handler/collectors/base.py                            85     12    86%   46, 67, 128-131, 142-145, 157-160, 165-167, 176-179
handler/collectors/base_test.py                       50      0   100%
handler/collectors/literal_collector.py               17      6    65%   25-27, 33-35
handler/collectors/literal_collector_test.py          20      0   100%
handler/collectors/perfzero_collector.py              29      6    79%   32-33, 52-57
handler/collectors/perfzero_collector_test.py         23      0   100%
handler/collectors/tensorboard_collector.py           78      9    88%   68-72, 90-91, 107-108, 124
handler/collectors/tensorboard_collector_test.py      50      0   100%
handler/utils.py                                      23      0   100%
handler/utils_test.py                                 11      0   100%
metrics_pb2.py                                       140      0   100%
--------------------------------------------------------------------------------
TOTAL                                                715     82    89%
```

That's baseline coverage for the current metrics handler is 67%.

File: metrics_v2/handler/collectors/__init__.py [hunk 12]
```diff
@@ -0,0 +1,3 @@
+from . import  literal_collector as literal
```
Comment (zcain117) [inline metrics_v2/handler/collectors/__init__.py:1]: nit extra space after import

File: metrics_v2/handler/alerts.py [hunk 13]
```diff
@@ -0,0 +1,77 @@
...
+import collections
+import datetime
+import logging
+import pytz
+import typing
+
+import jinja2
+from sendgrid.helpers import mail
+
+# Make sure these match the names of your Cloud Secrets. Refer to the README
```
Comment (zcain117) [inline metrics_v2/handler/alerts.py:24]: Not sure what the word "these" is referring to

Comment (zcain117) [reply metrics_v2/handler/alerts.py:24]: also I think this README doesn't exist yet

File: metrics_v2/handler/main.py [hunk 14]
```diff
@@ -0,0 +1,192 @@
...
+SEND_EMAIL_ALERTS = os.getenv('SEND_EMAIL_ALERTS', None)
+PROJECT = os.getenv('GCP_PROJECT', None)
+
+SOURCE_TO_COLLECTOR = {
+  'literals': collectors.literal.LiteralCollector,
+  'perfzero': collectors.perfzero.PerfZeroCollector,
+  'tensorboard': collectors.tensorboard.TensorBoardCollector,
+}
+
+_SENDGRID_API_SECRET_NAME = 'sendgrid-api-key'
```
Comment (zcain117) [inline metrics_v2/handler/main.py:47]: maybe this is where the other comment about secrets was meant to be?

**Review by zcain117 — APPROVED** [review]

File: metrics_v2/handler/main.py (hunk 14 above)
Comment (will-cromar) [reply metrics_v2/handler/main.py:49]: Yes, I moved these constants from alerts and missed the comment.

File: metrics_v2/handler/alerts.py (hunk 13 above)
Comment (will-cromar) [reply metrics_v2/handler/alerts.py]: I will copy over the instructions here: https://github.com/GoogleCloudPlatform/ml-testing-accelerators/tree/master/metrics_handler#optional-set-up-notification-emails

File: metrics_v2/handler/collectors/__init__.py (hunk 12 above)
Comment (will-cromar) [reply metrics_v2/handler/collectors/__init__.py]: Fixed.
